File: src/core/stdlib/maths/elementary.ts
/**
 * Elementary functions. Hypotenuse and exp/log use sections 20.10.1 and
 * 20.10.2. Square root uses its correctly rounded host operation. Power and
 * trigonometry still call host approximations under gap 1.
 */
import type { Value } from '../values/index.js';
import { NONE, isPresent, result } from '../values/index.js';
import { hypotenuse } from './hypot.js';
import { transcendental } from './transcendental.js';

/**
 * `sqrt(x)`: square root, absent below zero.
 *
 * Correctly rounded on every conforming platform, so this one is bit-identical
 * everywhere without a portable implementation of its own.
 */
export function sqrt(x: Value): Value {
  if (!isPresent(x) || x < 0) return NONE;
  return result(Math.sqrt(x));
}

/** `pow(x, y)`: absent where the result is not a finite real. */
export function pow(x: Value, y: Value): Value {
  if (!isPresent(x) || !isPresent(y)) return NONE;
  return result(Math.pow(x, y));
}

/** `exp(x)`: e to the power x. */
export function exp(x: Value): Value {
  return isPresent(x) ? transcendental('exp', x) : NONE;
}

/** `log(x)`: natural logarithm, absent at or below zero. */
export function log(x: Value): Value {
  if (!isPresent(x) || x <= 0) return NONE;
  return transcendental('log', x);
}

/** `log10(x)`: base ten logarithm, absent at or below zero. */
export function log10(x: Value): Value {
  if (!isPresent(x) || x <= 0) return NONE;
  return transcendental('log10', x);
}

/** `math.log2(x)`: base two logarithm, absent at or below zero. */
export function log2(x: Value): Value {
  if (!isPresent(x) || x <= 0) return NONE;
  return transcendental('log2', x);
}

/** The circle constant. */
export const PI: number = Math.PI;

/** The base of the natural logarithm. */
export const E: number = Math.E;

/** `math.hypot(x, y)`: `sqrt(x * x + y * y)` without intermediate overflow. */
export function hypot(x: Value, y: Value): Value {
  if (!isPresent(x) || !isPresent(y)) return NONE;
  return hypotenuse(x, y);
}

/** `math.toDegrees(x)`: radians to degrees. */
export function toDegrees(x: Value): Value {
  return isPresent(x) ? result((x * 180) / Math.PI) : NONE;
}

/** `math.toRadians(x)`: degrees to radians. */
export function toRadians(x: Value): Value {
  return isPresent(x) ? result((x * Math.PI) / 180) : NONE;
}

/** `math.sin(x)`: sine of an angle in radians. */
export function sin(x: Value): Value {
  return isPresent(x) ? result(Math.sin(x)) : NONE;
}

/** `math.cos(x)`: cosine. */
export function cos(x: Value): Value {
  return isPresent(x) ? result(Math.cos(x)) : NONE;
}

/** `math.tan(x)`: tangent. */
export function tan(x: Value): Value {
  return isPresent(x) ? result(Math.tan(x)) : NONE;
}

/** `math.asin(x)`: inverse sine, absent outside -1 to 1. */
export function asin(x: Value): Value {
  if (!isPresent(x) || x < -1 || x > 1) return NONE;
  return result(Math.asin(x));
}

/** `math.acos(x)`: inverse cosine, absent outside -1 to 1. */
export function acos(x: Value): Value {
  if (!isPresent(x) || x < -1 || x > 1) return NONE;
  return result(Math.acos(x));
}

/** `math.atan(x)`: inverse tangent. */
export function atan(x: Value): Value {
  return isPresent(x) ? result(Math.atan(x)) : NONE;
}

/** `math.atan2(y, x)`: angle of a vector, correct in all four quadrants. */
export function atan2(y: Value, x: Value): Value {
  if (!isPresent(y) || !isPresent(x)) return NONE;
  return result(Math.atan2(y, x));
}

File: engine/openscript/library/elementary.py
"""Elementary numeric functions with reproducible rounding.

Hypotenuse and exp/log use the exact integer recipes in sections 20.10.1 and
20.10.2; power and trigonometry use sections 20.10.3 and 20.10.4.
Square root uses its correctly rounded host operation.
"""

import math

from .hypotenuse import hypotenuse
from .transcendental import transcendental
from .power import power as real_power
from .trigonometric import trigonometric
from .values import ABSENT, Value, number, result

# The circle constant and the base of the natural logarithm, as `stdlib.md`
# section 8.2 names them. Both are the nearest binary64 to the constant and are
# the same bits on every platform.
PI: float = math.pi
E: float = math.e


def sqrt(x: Value) -> Value:
    """``sqrt(x)``: square root, absent below zero."""
    value = number(x)
    if value is None or value < 0:
        return ABSENT
    return result(math.sqrt(value))


def exp(x: Value) -> Value:
    """``exp(x)``: e to the power x, using section 20.10.2."""
    value = number(x)
    return ABSENT if value is None else transcendental('exp', value)


def log(x: Value) -> Value:
    """``log(x)``: portable natural logarithm, absent at or below zero."""
    value = number(x)
    if value is None or value <= 0:
        return ABSENT
    return transcendental('log', value)


def log10(x: Value) -> Value:
    """``log10(x)``: portable base ten logarithm, same absence rule."""
    value = number(x)
    if value is None or value <= 0:
        return ABSENT
    return transcendental('log10', value)


def log2(x: Value) -> Value:
    """``math.log2(x)``: portable base two logarithm, same absence rule."""
    value = number(x)
    if value is None or value <= 0:
        return ABSENT
    return transcendental('log2', value)


def power(x: Value, y: Value) -> Value:
    """``pow(x, y)``: portable real power, using section 20.10.3."""
    base = number(x)
    exponent = number(y)
    if base is None or exponent is None:
        return ABSENT
    return real_power(base, exponent)


def hypot(x: Value, y: Value) -> Value:
    """``math.hypot(x, y)``: the correctly rounded exact diagonal, section 20.10.1."""
    left = number(x)
    right = number(y)
    if left is None or right is None:
        return ABSENT
    return hypotenuse(left, right)


def to_degrees(x: Value) -> Value:
    """``math.toDegrees(x)``: ``(x * 180) / pi``, multiply first.

    Not a gap: it is two operations over binary64 and `stdlib.md` section 20.7
    fixes their order. Folding the constant into one factor is a different number
    at about a quarter of the arguments, so the association is written out rather
    than handed to a conversion the host provides.
    """
    value = number(x)
    return ABSENT if value is None else result((value * 180) / math.pi)


def to_radians(x: Value) -> Value:
    """``math.toRadians(x)``: ``(x * pi) / 180``, multiply first. Not a gap."""
    value = number(x)
    return ABSENT if value is None else result((value * math.pi) / 180)


def sin(x: Value) -> Value:
    """``math.sin(x)``: sine of an angle in radians, using section 20.10.4."""
    value = number(x)
    return ABSENT if value is None else trigonometric('sin', value)


def cos(x: Value) -> Value:
    """``math.cos(x)``: cosine, using section 20.10.4."""
    value = number(x)
    return ABSENT if value is None else trigonometric('cos', value)


def tan(x: Value) -> Value:
    """``math.tan(x)``: tangent, using section 20.10.4."""
    value = number(x)
    return ABSENT if value is None else trigonometric('tan', value)


def asin(x: Value) -> Value:
    """``math.asin(x)``: inverse sine, absent outside -1 to 1."""
    value = number(x)
    if value is None or value < -1 or value > 1:
        return ABSENT
    return trigonometric('asin', value)


def acos(x: Value) -> Value:
    """``math.acos(x)``: inverse cosine, same range rule."""
    value = number(x)
    if value is None or value < -1 or value > 1:
        return ABSENT
    return trigonometric('acos', value)


def atan(x: Value) -> Value:
    """``math.atan(x)``: inverse tangent, using section 20.10.4."""
    value = number(x)
    return ABSENT if value is None else trigonometric('atan', value)


def atan2(y: Value, x: Value) -> Value:
    """``math.atan2(y, x)``: angle of a vector, in all four quadrants.

    The vertical component comes first, as section 20.10.4 specifies.
    """
    vertical = number(y)
    horizontal = number(x)
    if vertical is None or horizontal is None:
        return ABSENT
    return trigonometric('atan2', vertical, horizontal)
